### PR TITLE
Metrics for layout structure comparison

### PR DESCRIPTION
**Pull request by Keerthana786**
- Added a **bleu score** metric collection for layout comparisons
-  Overall **accuracy** for the new test images set = **87.86%**

### CODE REVIEW

File: source/pic2card/commands/metrics/configs.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
+"""
+Gloabal settings and constants for metric collection.
+"""
+TEXTBOX = ["type", "text", "size", "color", "weight", "wrap"]
+ACTIONSET = ["type", "actions", "spacing"]
```
Comment (haridas) [inline source/pic2card/commands/metrics/configs.py:5]: I think key this as simple key mapping `dict`, so that generically you can do the sorting type independent fashion.

Comment (haridas) [reply source/pic2card/commands/metrics/configs.py:5]: Also for this only don't need to keep another file, you can load it as one dict variable and use it.

File: source/pic2card/commands/metrics/generate_bleu_score.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+            test = collections.OrderedDict(sorted(test.items()))
+            test["version"] = "1.3"
+            train_corpus = json.dumps(train)
+            train_corpus = re.sub(r"[^\w\s\{\}]", "", train_corpus)
+            test_corpus = json.dumps(test)
+            test_corpus = re.sub(r"[^\w\s\{\}]", "", test_corpus)
+
+            # Get metric
+            smoothing = translate.bleu_score.SmoothingFunction().method7
+            score = sentence_bleu([train_corpus.lower().split()],
```
Comment (haridas) [inline source/pic2card/commands/metrics/generate_bleu_score.py:166]: try corpus bleu ?

Comment (haridas) [reply source/pic2card/commands/metrics/generate_bleu_score.py:166]: in our case, both might behave similar way, but take a look

File: source/pic2card/commands/metrics/generate_bleu_score.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+                                  weights=(0.5, 0.5))
+            if score > 1:
+                smoothing = translate.bleu_score.SmoothingFunction().method3
+                score = sentence_bleu([train_corpus.lower().split()],
+                                      test_corpus.lower().split(),
+                                      smoothing_function=smoothing,
+                                      weights=(0.5, 0.5))
+
+            print(f"Similarity score for {train_image} is : {score}\n")
+            avg_score += score
```
Comment (haridas) [inline source/pic2card/commands/metrics/generate_bleu_score.py:178]: Pay be collect per image metric in a pandas data frame and export as CSV, just for the analysis purpose. May be debug mode we can dump csv, processed text of both test and train. To review things if necessary in the future.

File: source/pic2card/commands/metrics/metrics_helper.py [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+import re
+from datetime import datetime
+from typing import List, Dict
+
+import configs
+
+
+class Helper:
+    """
+    This is a general helper class for the metric pre-processing module.
```
Comment (haridas) [inline source/pic2card/commands/metrics/metrics_helper.py:15]: stress on actually what are those specific pre-processing, and high-level ideas what this class helps.

File: source/pic2card/commands/metrics/metrics_helper.py [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+        @return: date string in words format
+        """
+        date_string = \
+            re.findall(r"\b\d+\-\d+-\d+T\d+:\d+:\d+\w{1}", expression)[0]
+        date_time_obj = datetime.strptime(date_string, '%Y-%m-%dT%H:%M:%SZ')
+        date_string = date_time_obj.strftime("%a, %d %b %Y")
+        return date_string
+
+    def factset_to_textbox(self, element: Dict) -> List:
+        """
```
Comment (haridas) [inline source/pic2card/commands/metrics/metrics_helper.py:31]: is this transformation was done for ground truth also ?, we have FactSet aggregation in our pipeline right? so why this change needed for both the train and test dataset?

File: source/pic2card/commands/metrics/metrics_helper.py [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+        return element
+
+    def actions(self, element: Dict) -> List:
+        """
+        Performs all the pre-process operation related to actions inside an
+        Action-set
+        @param element: adaptive cards design element
+        @return: pre-processed list of action elements
+        """
+        element = self.remove_fields_from_element(element[0],
```
Comment (haridas) [inline source/pic2card/commands/metrics/metrics_helper.py:118]: all these methods can be absorbed in a single method and driven by the dict based keys or some sort of common logic can be done here?  I think we can absorb it, take a look so that we can avoid multiple helper classes that kinda have similar methods, and doing logics that can be merged into one common method for all.

Have the configuration dict, as mentioned above as a place to check which keys are required for which type of an element also this is the case for which keys has to be removed ( for this also it's common for all element type right ? )

File: source/pic2card/commands/metrics/README.MD [hunk 7]
```diff
@@ -0,0 +1,17 @@
+## Thins to pre-process manually while collecting the ground truth adaptive card json
+ 1. Filter out the elements not supported by pic2card except factset elements 
+    - The supported elements are: **TextBlock, ChoiceSet, ImageSet, CheckBox, 
+                                    ColumnSet, Image, ActionSet**
+ 2. Update the action elements as separate set of ActionSets with Submit action.
+ 3. Update the un-supported column width properties like **Weighted** and 
+    **Pixels** to supported formats ( **Stretch**, **Automatic** )
+ 4. Template data binding is not supported in metric collection.
+
```
Comment (haridas) [inline source/pic2card/commands/metrics/README.MD:9]: maybe mention for the Ground truth what are the changes done, and for the train set or predicted result what are the changes required before comparing it each other, and some reasoning about why we do this ( may be to make an apple to apple comparison or so :) )

File: source/pic2card/commands/metrics/metrics_helper.py [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+        Fetch the design template for the column-set element and calls the
+        needed pre-processing functions.
+        @param body: design element's layout structure
+        @param element: adaptive card design element
+        """
+        backup = copy.deepcopy(element)
+        template_object = getattr(self.helper_object,
+                                  element.get("type", ""))
+        element = template_object(element)
+        element = collections.OrderedDict(sorted(element.items()))
```
Comment (haridas) [inline source/pic2card/commands/metrics/metrics_helper.py:279]: python2.7 onwards the keys in a dict keep the insertion order. So explicit OrderedDict thing isn't required. This happened because of the introduction of new dictionary implementation in python.

File: source/pic2card/commands/metrics/metrics_helper.py (hunk 5 above)
Comment (Keerthana786) [reply source/pic2card/commands/metrics/metrics_helper.py:81]: We do not support FactSet element in our pipeline currently, right now all the FactSet key-value elements will be detected as rows-columns of textboxes.

File: source/pic2card/commands/metrics/generate_bleu_score.py [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+                                    [test_corpus.lower().split()],
+                                    smoothing_function=smoothing,
+                                    weights=(0.5, 0.5))
+
+            print(f"Similarity score for {train_image} is : {score}\n")
+            avg_score += score
+            if debug:
+                rows.append([train_image, score])
+                file_name = train_image.split("/")[-1]
+                if not os.path.exists('../../../ground_truth_dump'):
```
Comment (haridas) [inline source/pic2card/commands/metrics/generate_bleu_score.py:165]: Use some global variables to avoid constants throughout the code.

File: source/pic2card/commands/metrics/metrics_helper.py (hunk 5 above)
Comment (haridas) [reply source/pic2card/commands/metrics/metrics_helper.py:81]: We can add separate aggregation for this in our pipeline.. as factset is one that comes as part of the rowGroup identification time.

File: source/pic2card/commands/metrics/generate_bleu_score.py [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+            test = dict(sorted(test.items(), key=lambda s: s[0]))
+            test["version"] = "1.3"
+            train_corpus = json.dumps(train)
+            train_corpus = re.sub(r"[^\w\s\{\}]", "", train_corpus)
+            test_corpus = json.dumps(test)
+            test_corpus = re.sub(r"[^\w\s\{\}]", "", test_corpus)
+
+            # Get metric
+            smoothing = translate.bleu_score.SmoothingFunction().method7
+            score = corpus_bleu([[train_corpus.lower().split()]],
```
Comment (haridas) [inline source/pic2card/commands/metrics/generate_bleu_score.py:149]: Usage of corpus_bleu increased the overall score?

File: source/pic2card/commands/metrics/README.MD [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+        
+      - **Pixels**
+        
+      update it to supported values:  **Stretch** / **Automatic**. 
+ 4. Template data binding is not supported in metric collection.
+  
+    - Since we are comparing the layout structure and each element's 
+      property values, comparing a custom expression binding variables on one side
+      and expression binding done by pic2card / non-template binding
+      card JSON values, will be meaningless and lowers our layout metric score.
```
Comment (haridas) [inline source/pic2card/commands/metrics/README.MD:28]: These workarounds we can apply to our card generation pipeline to make it as close to the spec as possible. Let's evaluate the same.

File: source/pic2card/commands/metrics/generate_bleu_score.py (hunk 10 above)
Comment (Keerthana786) [reply source/pic2card/commands/metrics/generate_bleu_score.py:149]: With our data-set both **corpus_bleu** and **sentence_bleu** overall scores are same , **87.86%**

**Review by haridas — APPROVED** [review]
